File: cpp/34_digit_factorials.cpp
#include <iostream>
#include <vector>

int digit_factorials(){
    std::vector<int> factorials = {1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880};
    // when 9! * d < 10^(d - 1), d-digit number n will always be 
    // greater than the sum of the factorials of its digits
    // 362880 * d < 10^(d - 1) when d > 7, so n can have at most 7 digits
    int total = 0;
    for (int i = 3; i < 7 * 362880; i++){
        int factSum = 0, n = i;
        while (n > 0){
            factSum += factorials[n % 10];
            n /= 10;
        }
        if (factSum == i) total += i;
    }
    return total;
}

int main() {
    std::cout << digit_factorials() << "\n";
}

File: c#/_34_DigitFactorials.cs
using System;

public class _34_DigitFactorials{
	public static int Euler34(){
		int[] fact = new int[]{1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880};
		// an n-digit number cannot be a digit factorial sum
		// if the maximum n-digit factorial sum, n * 9! = 362880n
		// is less than the smallest n-digit number, 10^(n-1)
		// 362880n < 10^(n - 1) is true when n > 7
		// bounds to check: [10, 2540160]
		int res = 0;
		for (int n = 10; n < 7 * fact[9]; n++){
			int i = n, sum = 0;
			while (i > 0){
				sum += fact[i % 10];
				i /= 10;
			}
			if (sum == n) res += n;
		}
		return res;
	}
	
	public static void Main(){
		Console.WriteLine(Euler34());
	}
}
